specification/dsdl/versioning.tex

```diff
@@ -0,0 +1,411 @@
+\end{remark}
+
+\subsection{Versioning}
+
+\subsubsection{General assumptions}
+
+The concept of versioning applies only to composite data types.
+As such, unless specifically stated otherwise, every reference to ``data type''
+in this section implies a composite data type.
+
+A data type is uniquely identified by its full name,
+assuming that every root namespace is uniquely named.
+There is one or more versions of every data type.
+
+A data type definition is uniquely identified by its full name and the version number pair.
+In other words, there may be multiple definitions of a data type differentiated by their version numbers.
+
+\subsubsection{Versioning principles}
+
+Every data type definition has a pair of version numbers ---
+a major version number and a minor version number, following the principles of semantic versioning.
+
+For the purposes of the following definitions, a \emph{release} of a data type definition stands for
+the disclosure of the data type definition to its intended users or to the general public,
+or for the commencement of usage of the data type definition in a production system.
+
+In order to ensure a deterministic application behavior and ensure a robust migration path
+as data type definitions evolve, UAVCAN requires that all data type definitions that share the same
+full name and the same major version number must be semantically compatible with each other
+and mutually bit-compatible with each other.
+
+The versioning rules do not extend to scenarios where the name of a data type is changed,
+because that would essentially construe the release of a new data type,
+which relieves its designer from all compatibility requirements.
+When a new data type is first released,
+the version numbers of its first definition must be assigned ``1.0'' (major 1, minor 0).
+
+In order to ensure predictability and functional safety of applications that leverage UAVCAN,
+the standard requires that once a data type definition is released,
+its DSDL source text, name, version numbers, fixed port ID, and other properties cannot undergo any
+modifications whatsoever, with the following exceptions:
+\begin{itemize}
+    \item Whitespace changes of the DSDL source text are allowed,
+    excepting string literals and other semantically sensitive contexts.
+
+    \item Comment changes of the DSDL source text are allowed as long as such changes
+    do not affect semantic compatibility of the definition.
+
+    \item A deprecation marker directive (section~\ref{sec:dsdl_directives}) can be added or removed\footnote{%
+    Removal is useful when a decision to deprecate a data type definition is withdrawn.}.
+\end{itemize}
+Addition or removal of the fixed port identifier is not permitted after a data type definition
+of a particular version is released.
+
+Therefore, substantial modifications of released data types are only possible by releasing
```

Comment:
It's really hard to figure out what's being proposed the way it's worded in this section. We say "...the standard
requires that once a data type definition is released, its DSDL source text, name, version numbers, fixed
port ID, and other properties cannot undergo any modifications whatsoever" And then we say "...substantial modifications... are ...possible". These two statements contradict. I think we mean, in the previous paragraph, that you can only change a released type by releasing a new version of the type, correct?

Reply:
Fixed here: https://github.com/UAVCAN/specification/pull/46/commits/b58830f1ee8614d29007daf88525faf7cd538116

The original wording seemed technically correct, but it could be hard to follow, indeed. It said "substantial modifications of released data types", but you seem to have read it as "substantial modifications of released data type definitions". A definition is virtually unchangeable, whereas a "data type" is a set of its definitions of different versions, as implied under "General assumptions".